presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java

```diff
@@ -133,51 +135,52 @@ private static SqlScalarFunctionBuilder castFunctionToDecimalFromBuilder(String
                         .withExtraParameters((context) -> {
                             DecimalType resultType = checkType(context.getReturnType(), DecimalType.class, "resultType");
                             long tenToScale = longTenToNth(resultType.getScale());
-                            return ImmutableList.of(resultType.getPrecision(), resultType.getScale(), tenToScale);
+                            BigInteger tenToScaleLong = bigIntegerTenToNth(resultType.getScale());
+                            return ImmutableList.of(resultType.getPrecision(), resultType.getScale(), tenToScale, tenToScaleLong);
                         })
                 );
     }
 
     private DecimalCasts() {}
 
     @UsedByGeneratedCode
-    public static boolean shortDecimalToBoolean(long decimal, long precision, long scale, long tenToScale)
+    public static boolean shortDecimalToBoolean(long decimal, long precision, long scale, long tenToScaleShort, BigInteger tenToScaleLong)
     {
         return decimal != 0;
     }
 
     @UsedByGeneratedCode
-    public static boolean longDecimalToBoolean(Slice decimal, long precision, long scale, long tenToScale)
+    public static boolean longDecimalToBoolean(Slice decimal, long precision, long scale, long tenToScaleShort, BigInteger tenToScaleLong)
     {
         return !decodeUnscaledValue(decimal).equals(ZERO);
     }
 
     @UsedByGeneratedCode
-    public static long booleanToShortDecimal(boolean value, long precision, long scale, long tenToScale)
+    public static long booleanToShortDecimal(boolean value, long precision, long scale, long tenToScaleShort, BigInteger tenToScaleLong)
     {
-        return value ? tenToScale : 0;
+        return value ? tenToScaleShort : 0;
     }
 
     @UsedByGeneratedCode
-    public static Slice booleanToLongDecimal(boolean value, long precision, long scale, long tenToScale)
+    public static Slice booleanToLongDecimal(boolean value, long precision, long scale, long tenToScaleShort, BigInteger tenToScaleLong)
     {
-        return encodeUnscaledValue(BigInteger.valueOf(value ? tenToScale : 0L));
+        return encodeUnscaledValue(value ? tenToScaleLong : ZERO);
     }
 
     @UsedByGeneratedCode
-    public static long shortDecimalToBigint(long decimal, long precision, long scale, long tenToScale)
+    public static long shortDecimalToBigint(long decimal, long precision, long scale, long tenToScaleShort, BigInteger tenToScaleLong)
```

Comment:
looking at those methods I'm getting lots of `WTF?!` errors. What's the difference between `tenToScaleShort` and `tenToScaleLong`? What does it mean short and long? When should I use long and when short version? Are they mutually exclusive?

Is this `tenToScaleShort` prone to overflow with `scale > 18`? I think there should be only one of this parameters available and each of this function should always use/get only one - not both params.